src/claim_proofs/ffi_wrapper/src/lib.rs

```diff
@@ -0,0 +1,271 @@
+
+extern crate libc;
+use libc::size_t;
+use std::slice;
+
+use cryptography::claim_proofs::{build_scope_claim_proof_data, compute_cdd_id, compute_scope_id};
+
+pub type ScopeClaimData = cryptography::claim_proofs::ScopeClaimData;
+pub type ScopeClaimProofData = cryptography::claim_proofs::ScopeClaimProofData;
+pub type ProofPublicKey = cryptography::claim_proofs::ProofPublicKey;
+pub type ProofKeyPair = cryptography::claim_proofs::ProofKeyPair;
+pub type CDDClaimData = cryptography::claim_proofs::CDDClaimData;
+pub type Signature = schnorrkel::Signature;
+pub type Scalar = curve25519_dalek::scalar::Scalar;
+pub type RistrettoPoint = curve25519_dalek::ristretto::RistrettoPoint;
+
+fn box_alloc<T>(x: T) -> *mut T {
+    Box::into_raw(Box::new(x))
+}
+
+// ------------------------------------------------------------------------
+// Data Structures
+// ------------------------------------------------------------------------
+
+/// Create a `Scalar` object from a 32 bytes array.
+///
+/// Caller is responsible for calling `scalar_free()` to deallocate this object.
+/// SAFETY: Caller is also responsible for making sure the `scalar_bits` pointer
+///         is not Null and points to a 32 bytes block of memory.
+#[no_mangle]
+pub unsafe extern "C" fn scalar_new(scalar_bits: *const u8) -> *mut Scalar {
+    assert!(!scalar_bits.is_null());
+
+    let len = 32usize;
+    let mut scalar_slice = [0u8; 32];
+    scalar_slice.copy_from_slice(&slice::from_raw_parts(scalar_bits, len)[..32]);
+
+    let scalar = Scalar::from_bits(scalar_slice);
+    box_alloc(scalar)
+}
+
+/// Deallocates a `Scalar` object's memory.
+#[no_mangle]
+pub unsafe extern "C" fn scalar_free(ptr: *mut Scalar) {
+    if ptr.is_null() {
+        return;
+    }
+
+    Box::from_raw(ptr);
+}
+
+/// Create a new `CDDClaimData` object.
+///
+/// Caller is responsible for calling `cdd_claim_data_free()` to deallocate this object.
+/// SAFETY: Caller is also responsible for making sure the `investor_did` and
+///         `investor_unique_id` pointers are not Null.
+#[no_mangle]
+pub unsafe extern "C" fn cdd_claim_data_new(
+    investor_did: *mut Scalar,
+    investor_unique_id: *mut Scalar,
+) -> *mut CDDClaimData {
+    assert!(!investor_did.is_null());
+    assert!(!investor_unique_id.is_null());
+
+    let investor_did: Scalar = *investor_did;
+    let investor_unique_id: Scalar = *investor_unique_id;
+    box_alloc(CDDClaimData {
+        investor_did,
+        investor_unique_id,
+    })
+}
+
+/// Deallocates a `CDDClaimData` object's memory.
+#[no_mangle]
+pub unsafe extern "C" fn cdd_claim_data_free(ptr: *mut CDDClaimData) {
+    if ptr.is_null() {
+        return;
+    }
+    Box::from_raw(ptr);
+}
+
+/// Create a new `ScopeClaimData` object.
+///
+/// Caller is responsible for calling `scope_claim_data_free()` to deallocate this object.
+/// SAFETY: Caller is also responsible for making sure the `scope_did` and
+///         `investor_unique_id` pointers are not Null.
```

Comment:
Not sufficient; they have to point to valid & allocated `Scalar`s too (applies in other functions two where you've noted this about Null).

Reply:
I think the function signature is quite clear that these arguments are `Scalar`...

Reply:
The "SAFETY" comment here states that the requirement is `!= ptr::null()`, but this is sufficient, so I think that's misleading the reader to think this is sufficient when it's not. It's also not obvious (although probable) that this could be any pointer to a valid & allocated `Scalar`. It might for example be one that needs to be heap allocated. So if we do have a SAFETY comment (and its good practice to have one), it's better to be exact (while not being verbose, it's not a Coq proof you need here...) than to leave things inferred.